# Estado de Bell o Hello World quântico do emaranhamento

### Instalação da lib QisKit

In [17]:
!python.exe -m pip install --upgrade pip
!pip install qiskit qiskit_aer matplotlib

  Using cached matplotlib-3.10.8-cp313-cp313-win_amd64.whl.metadata (52 kB)
  Using cached contourpy-1.3.3-cp313-cp313-win_amd64.whl.metadata (5.5 kB)
  Using cached cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)
  Using cached pillow-12.1.1-cp313-cp313-win_amd64.whl.metadata (9.0 kB)
  Using cached pyparsing-3.3.2-py3-none-any.whl.metadata (5.8 kB)
Using cached matplotlib-3.10.8-cp313-cp313-win_amd64.whl (8.1 MB)
Using cached contourpy-1.3.3-cp313-cp313-win_amd64.whl (226 kB)
Using cached cycler-0.12.1-py3-none-any.whl (8.3 kB)
   ---------------------------------------- 0.0/2.3 MB ? eta -:--:--
   -------- ------------------------------- 0.5/2.3 MB 3.5 MB/s eta 0:00:01
   ---------------------------------------- 2.3/2.3 MB 7.7 MB/s  0:00:00
Using cached pillow-12.1.1-cp313-cp313-win_amd64.whl (7.0 MB)
Using cached pyparsing-3.3.2-py3-none-any.whl (122 kB)

   ---------------------------------------- 0/7 [pyparsing]
   ---------------------------------------- 0/7 [pyparsing]
   ----

In [14]:
from qiskit import QuantumCircuit, transpile
from qiskit_aer import Aer

#### 1. Criando um circuito com 2 qubits e 2 bit clássicos

In [5]:
qc = QuantumCircuit(2, 2)

#### 2.Aplicando uma porta Hadmard (H) no qubit 0 para criar superposição
- Isso coloca o qubit 0 em superposição (|0> + |1>) / sqrt(2)

In [6]:
qc.h(0)

#### 3.Aplica uma porta CNOT (Cx) para emaranhar o qubit 0 (controle) com o qubit 1 (alvo)
- Control=qubit 0, Target=qubit 1. Isso gera emaranhamento.

In [7]:
qc.cx(0, 1)

#### 4.Realiza a medição dos qubits para os bits clássicos correspondentes
- Mapeia o qubit 0 no bit 0, e o qubit 1 no bit 1

In [12]:
qc.measure([0, 1], [0, 1])

#### 5. Define o backend e executa

In [15]:
backend = Aer.get_backend("qasm_simulator")
tqc = transpile(qc, backend)
job = backend.run(tqc, shots=1024) # Rodando o circuto 1024 vezes

#### 6.Resultados

In [16]:
result = job.result()
counts = result.get_counts()
print(f"Resultados (Contagens): {counts}")

Resultados (Contagens): {'11': 527, '00': 497}


##### O que o resultado quer dizer?
Nesse teste o circuito foi executado 1024 vezes
- '00': 497 -> Quase metade dos experimentos ambos os qubits colapsaram para o estado 0
- '11': 527 -> Na outra metade, ambos colapsaram para o estado 1
- Nenhum dos qubist foram encontrados em estados diferentes

Conclusão -> Os qubits estão correlacionados de forma que, se você medir um e descobrir que ele é 1, o outro será obrigatoriamente 1 também, instantanemente, não importa a distância entre eles

#### Circuito quântico desenhado

In [19]:
print(qc.draw(output='text'))

     ┌───┐     ┌─┐   
q_0: ┤ H ├──■──┤M├───
     └───┘┌─┴─┐└╥┘┌─┐
q_1: ─────┤ X ├─╫─┤M├
          └───┘ ║ └╥┘
c: 2/═══════════╩══╩═
                0  1 


- Porta H (Hadamard)
    - Transforma um estado definino em um estado de superposição, onde o qubit existe como uma combinação linear de ambos os estados simultaneamente
- Porta X (Pauli-X) ou Porta NOT
    - Realiza uma inversão de bit (bit flip)
- M (Medição/Measurement)
    - É o processo de observar o qubit para extrair informação quântica e convertê-la em bits clássicos (0 ou 1)